### PR TITLE
Added code sample according to bem code style

### PR DESCRIPTION
**Pull request by qfox**
Please take a look at code sample. Suggest changes if we need to describe more edge cases

To test or make a fix:

``` sh
git@github.com:bem/jscs-bem.git
cd jscs-bem
open test/presets/bem.js
jscs -c ./test/presets/.jscsrc test/presets/bem.js
```

/cc @veged @dfilatov @tadatuta @narqo

### CODE REVIEW

File: test/presets/bem.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
+/* jshint unused:false, undef:false */
+
+// Names
+var SOME_CONSTANT = 1;
+var foo = new Foo();
```
Comment (arikon) [inline test/presets/bem.js]: Один `var` на несколько деклараций же

File: test/presets/bem.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+// Multiple var declaration
+function testVars() {
+    var keys = ['foo', 'bar'];
```
Comment (arikon) [inline test/presets/bem.js]: Один `var` на несколько деклараций же

Comment (qfox) [inline test/presets/bem.js]: https://github.com/bem/bem-core/blob/v2/common.blocks/i-bem/i-bem.vanilla.js#L286 — разброд ;)

Comment (narqo) [inline test/presets/bem.js]: @zxqfox надо отличать:

``` js
var a = 1;
var b = 1;
var c = 1;
```

от

``` js
var a = 1,
    b = 1;

if(foo) {}
bar();

var c = 1;
```

В нашей формулировке правило звучит как-то так: «один `var` на блок идущих подряд деклараций». Фанатичной идеи «унести все декларации переменных в один var-блок» — нет.

Сомневаюсь, правда, что современные code-style-чекеры в состоянии проверить это. Так же как и отсутствие индента в корневом модуле-обертке :(

```
(function(exports, $, undefined) {

exports.Block = function() {     // ☚ индент бессмысленен
    ···                         // ☚ индент 4 пробела
};

}(window, jQuery))
```

Comment (qfox) [inline test/presets/bem.js]: > Сомневаюсь, правда, что современные code-style-чекеры в состоянии проверить это.

Умеют ;-) Они и не такое умеют.

> exports.Block = function() {     // ☚ индент бессмысленен

С этим не ковырялся, но тоже можно сделать. Надо собрать кейсы.

**upd** С `var` блоком и индентом понятно, все остальное ок?

Comment (qfox) [inline test/presets/bem.js]: Поправил `var`. Теперь ок?

**Comment by arikon** [comment]
Вроде да

**Comment by qfox** [comment]
@narqo @arikon What you think about that block with module/block declaration? https://github.com/bem/jscs-bem/commit/f414b7b9d2ceed496f39974199d6b9eb90f25f33

File: test/presets/bem.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+// Comments with url can have maximum line length > 120 ................... https://github.com/yandex/codestyle/blob/master/js.md
+
+// `i-bem` should be named BEM
+// `i-bem__dom` should be named BEMDOM
+modules.define('something', ['i-bem__dom'], function(provide, BEMDOM) {
+
+// Should use `this.name` if possible for BEMDOM.decl
+// Should have @lends jsdoc tag
+BEMDOM.decl(this.name, /** @lends Something.prototype */ {
```
Comment (arikon) [inline test/presets/bem.js]: Я только не понимаю, почему мы здесь отступ не добавляем

File: test/presets/bem.js [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+});
+
+[1, 2, 3].forEach(function(n) {
+    this.fn(n);
+}, this);
+
+// Comments with url can have maximum line length > 120 ................... https://github.com/yandex/codestyle/blob/master/js.md
+
+// `i-bem` should be named BEM
+// `i-bem__dom` should be named BEMDOM
```
Comment (arikon) [inline test/presets/bem.js]: Почему именно так?

File: test/presets/bem.js (hunk 3 above)
Comment (narqo) [inline test/presets/bem.js]: > почему мы здесь отступ не добавляем

Тогда основная масса кода будет написана с отступом в 8-пробелов. Потому что у нас каждый JS-код завернут в `modules.define()` — зачем? При средней ширине строки 80-символов (на мониторе в 13", в vim открыть два сплит-буфера или в Idea развернуть панель ФС), количество пустого места слева, нормально так затрудняем чтение кода (да это вкусовщина, но тут, кажется, как со всем code-style'ом)

> @lends Something.prototype

В `bem-components` мы описываем результат `BEMDOM.decl`, как:

``` js
/**
 * @class something    // ☚ lowercase + @bem :(
 * @bem
 */
```

"lowercase" потому что это в первую очередь блок: он привязан к CSS-классу "something", в нем может быть дефис (`control-group`) в названии. То что это еще и «ООП-класс» уходит на второй план. Поэтому везде `@lends something.prototype`. 

Переменные в define, называем `Something` — да это неоднозначность, но что делать с `control-group`, непонятно (CamelCase везде, включая CSS-классы, оказался слишком радикальной идеей).

File: test/presets/bem.js [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+        'disabled' : {
+            'true' : function() {
+            },
+            '' : function() {
+            }
+        }
+    }
+}, /** @lends Something */ {
+    live: true
```
Comment (narqo) [inline test/presets/bem.js]: `live : true` — пробем перед ":"

File: test/presets/bem.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+[1, 2, 3].forEach(function(n) {
+    this.fn(n);
+}, this);
+
+// Comments with url can have maximum line length > 120 ................... https://github.com/yandex/codestyle/blob/master/js.md
+
+// `i-bem` should be named BEM
+// `i-bem__dom` should be named BEMDOM
+modules.define('something', ['i-bem__dom'], function(provide, BEMDOM) {
```
Comment (narqo) [inline test/presets/bem.js]: А где вызов `provide`?

Comment (qfox) [inline test/presets/bem.js]: Резонно ;-)

File: test/presets/bem.js (hunk 5 above)
Comment (qfox) [inline test/presets/bem.js]: добавил

File: test/presets/bem.js (hunk 3 above)
Comment (qfox) [inline test/presets/bem.js]: Поправил `@lends`, добавил пример блока с минусом в имени в `define`.

**Comment by qfox** [comment]
@narqo @arikon can you take a look at it once more?

/cc @tadatuta @sipayRT

**Comment by sipayRT** [comment]
:+1:

File: test/presets/bem.js [hunk 7]
```diff
@@ -151,6 +151,43 @@ doAsync(function() {
...
+    beforeSetMod : {
+        'hovered' : {
+            'true' : function() {
+            }
+        }
+    },
+
+    // Should be line-break between mods declarations
+    onSetMod : {
+        // `any` (`*`) value should be used explicitly
```
Comment (narqo) [inline test/presets/bem.js]: Where this requirement is came from? In general, we use `*` only in case `other` mod handlers are present. [Here is some random example from `menu`](https://github.com/bem/bem-components/blob/e594e82f3a59f7a52f97b28e2363b12740e9e6fa/common.blocks/menu/menu.js#L36)

File: test/presets/bem.js [hunk 8]
```diff
@@ -151,6 +151,43 @@ doAsync(function() {
 
 // Comments with url can have maximum line length > 120 ................... https://github.com/yandex/codestyle/blob/master/js.md
 
+// `i-bem` should be named BEM
+// `i-bem__dom` should be named BEMDOM
+modules.define('something-truly', ['i-bem__dom', 'something-else'], function(provide, BEMDOM, SomethingElse) {
+
+// Should use `this.name` if possible for BEMDOM.decl
+// Should have @lends jsdoc tag
```
Comment (narqo) [inline test/presets/bem.js:159]: I'm not sure, but seems that `@lends` may be only used if `@class` tag declared. In other cases it won't work properly

Comment (qfox) [inline test/presets/bem.js:159]: Should we note that in comment? Should we add `@class`?

I don't see why It is required.

File: test/presets/bem.js (hunk 7 above)
Comment (qfox) [inline test/presets/bem.js]: Just my feelings. That's why I want a review.

So am I right that we shouldn't use `*` if it's only `*` (as it's done in example), but can use it `*` if there are other options?

```
'checked' : function (modName, modVal) { // as any
},
'rechecked' : {
  'concrete' : function () {},
  '*' : function () {} // as others?
}
```

But not

```
'checked' : {
  '*' : function (modName, modVal) {
  }
}
```

Comment (narqo) [inline test/presets/bem.js]: > So am I right that we shouldn't use `*` if it's only `*` (as it's done in example), but can use it `*` if there are other options?

Yes, this is the recommendation that we stick to in our code.

Comment (qfox) [inline test/presets/bem.js]: Thanks, need to fix this.

File: test/presets/bem.js (hunk 8 above)
Comment (narqo) [inline test/presets/bem.js:159]: > I don't see why It is required.

It seems I was wrong. [`class` plugin for `jsd`](https://github.com/bem/jsd-plugins-bem/blob/08080237503b31b2f701110dcc4707071ead43d0/plugins/class.js) should work properly. At least the new `class` entity will be parsed from the tag.

**Comment by qfox** [comment]
I'm gonna to merge this tomorrow.

**Comment by qfox** [comment]
Tomorrow is NOW. lel.
